code/krepel/math/vector3.d

```diff
@@ -0,0 +1,569 @@
+module krepel.math.vector3;
+
+import krepel.math.math;
+import krepel.algorithm.comparison;
+import krepel.math.vector2;
+import krepel.math.vector4;
+import std.conv;
+
+@nogc:
+@safe:
+nothrow:
+
+/// Calculates the Dot Product of the two Vectors
+/// Input Vectors will not be modified
+float Dot(Vector3 Lhs, Vector3 Rhs)
+{
+  float Result = Lhs | Rhs;
+  return Result;
+}
+
+/// Multiplies two Vectors per component returning a new Vector containing
+/// (Lhs.X * Rhs.X, Lhs.Y * Rhs.Y, Lhs.Z * Rhs.Z)
+/// Input Vectors will not be modified
+Vector3 Mul(Vector3 Lhs, Vector3 Rhs)
+{
+  return Lhs * Rhs;
+}
+
+/// Calculates the cross Product (Lhs x Rhs) and returns the result
+/// Input Vectors will not be modified
+Vector3 Cross(Vector3 Lhs, Vector3 Rhs)
+{
+  return Lhs ^ Rhs;
+}
+
+/// Calculates the squared length  of the given Vector, which is the same as the dot product with the same Vector
+/// Input Vector will not be modified
+float LengthSquared(Vector3 Vec)
+{
+  return Vec | Vec;
+}
+
+/// Calculates the squared length of the X and Y components, ignoring the Z component
+/// Input Vector will not be modified
+float LengthSquared2D(Vector3 Vec)
+{
+  return Vec.X * Vec.X + Vec.Y * Vec.Y;
```

Comment:
Also note that the here used simd instructions are far from optimal, but this is a task for the extra simd math.